src/member.ts

```diff
@@ -0,0 +1,358 @@
+  from: ID;
+  // Our response to this request
+  response: ResponseType;
+}
+
+export class RequestState {
+  req: Request;
+  signatures: {
+    [id in ID]: Response;
+  };
+  finished: boolean;
+
+  constructor(req: Request) {
+    this.req = req;
+    this.signatures = {};
+    this.finished = false;
+  }
+
+  getState(
+    neededSignatures: number,
+    maxDenied: number = neededSignatures
+  ): ResponseState {
+    if (this.finished) return "finished";
+    if (this.isCancelled()) return "cancelled";
+    if (this.isConflicted()) return "failed";
+    if (this.numberAccepted() >= neededSignatures) return "ready";
+    if (this.numberDenied() >= maxDenied) return "failed";
+
+    return "pending";
+  }
+
+  addResponse(author: ID, response: Response) {
+    this.signatures[author] = response;
+  }
+
+  get id() {
+    return this.req.id;
+  }
+
+  get from() {
+    return this.req.from;
+  }
+
+  get operation() {
+    return this.req.operation;
+  }
+
+  numberDenied(): number {
+    return Object.keys(this.signatures).filter((id: ID) => {
+      return this.signatures[id].response === "deny";
+    }).length;
+  }
+
+  numberAccepted(): number {
+    return Object.keys(this.signatures).filter((id: ID) => {
+      return this.signatures[id].response === "accept";
+    }).length;
+  }
+
+  isConflicted(): boolean {
+    return Object.keys(this.signatures).some((id: ID) => {
+      return this.signatures[id].response === "conflict";
+    });
+  }
+
+  isCancelled(): boolean {
+    return this.signatures[this.from]?.response === ("cancel" as ResponseType);
+  }
+
+  finish() {
+    this.finished = true;
+  }
+}
+
+function allowAll(request: Request): boolean {
+  return true;
+}
+
+export interface ShouldAcceptCB {
+  (request: Request): boolean;
+}
+
+export interface MemberConstructorOptions {
+  id?: ID;
+  initiator?: ID;
+  shouldAccept?: ShouldAcceptCB;
+}
+
+export class Member extends EventEmitter {
+  id: ID;
+  ownFeed: FeedItem[];
+  knownFeeds: {
+    [id in ID]: FeedItem[];
+  };
+  memberIndexes: {
+    [id in ID]: number;
+  };
+  ownIndex: number;
+  requests: {
+    [id in ID]: RequestState;
+  };
+  knownMembers: string[];
+  shouldAccept: ShouldAcceptCB;
+
+  constructor(
+    { id, initiator, shouldAccept = allowAll }: MemberConstructorOptions = {
+      shouldAccept: allowAll,
+    }
+  ) {
+    super();
+    this.id = id || makeID();
+    this.ownFeed = [];
+    this.ownIndex = 0;
+    this.knownFeeds = {
+      [this.id]: this.ownFeed,
+    };
+    this.memberIndexes = {
+      [this.id]: 0,
+    };
+    this.requests = {};
+    this.knownMembers = [initiator || this.id];
+    this.shouldAccept = shouldAccept;
+  }
+
+  sync(member: Member) {
+    // Iterate through member's knownFeeds
+    // Get any new data you don't have
+    // Get new data from members to knownFeeds
+    for (const id of Object.keys(member.knownFeeds)) {
+      const feed = member.knownFeeds[id];
+      const ownCopy = this.getFeedFor(id);
```

Comment:
There should be a validate template here to validate if the feed data that is added keeps being added.

Reply:
Would this require the signvector stuff?

Reply:
The sign-vectors are there too. Yes but in a simpler case: if there is a feed of a member that has been removed or was never added.

Reply:
Hmm, mind pseudocodeing how the verification would look and what the test for forking would look like?

Reply:
```javascript
function isIncomingFeedForked (localFeed, feed) {
  // Look at the merkle tree and see if the feed was forked at a previous step
}

function isIncomingFeedSignVectorCorrect (feed, startIndex) {
  let vector = createSignVectorFor(feed.get(startIndex))
  for (let i = startIndex+1; i<feed.length; i++) {
    if (!signVector.validate(feed.get(i))) return false
  }
  return true
}

function isIncomingFeedCorrect (feed) {
  // testing "correct-ness" of requests, i.e. proper msgpack/protobuf object; size limit not exceeded; removes feed that actually existed. Does the signature exist for the request? etc.
}

function verifyFeed (localFeed, feed, startIndex, peer) {
  if (isIncomingFeedForked(localFeed, feed)) {
    // Red alert
    requestRemoveMember({ cause: 'fork-detected' })
    return false
  }
  if (!isIncomingFeedSignVectorCorrect(feed, startIndex)) {
    addToBlackList(peer)
    return false
  }
  if(!isIncomingFeedCorrect(feed)) {
    // orange alert
    requestRemoveMember({ cause: 'fork-detected' })
    return false
  }
  return true
}
```

Reply:
Sweet, thank you!